### DISCUSSION TITLE
How to keep the commits signed with GPG key

### DISCUSSION THREAD

**Discussion by ShadyMedic** [newren/git-filter-repo#209] [Q&A] [answered]
Hi there.

I'm using git-filter-repo to purge secrets that I accidentally committed a long time ago (SMTP and database passwords) and everything works great, but I noticed that after I perform the operation, all signed commits aren't signed anymore.

Is it possible to tell the program to use a certain GPG key while rewriting the history of my repository?

Thanks in advance.

**Comment by newren** [comment, marked as answer]
git filter-repo builds on top of fast-export and fast-import.  fast-export has to strip the existing signatures because they won't be valid after the modifications to history.  And fast-import has no option to add a signature to commits with some specified key.  Such a feature would need to be added to upstream git, then filter-repo would be able to use it.

Sorry there isn't a better answer at this time.

**Reply by ShadyMedic** [reply]
Ok, thanks for letting me know.

**Reply by Hazimehhassan** [reply]
https://github.com/newren/git-filter-repo/issues/713#issue-3545372832

**Comment by vhscom** [comment]
Although you cannot keep the existing GPG signed commits you may always [resign them](https://stackoverflow.com/a/71350955/712334).

**Reply by heygarrett** [reply]
Just a heads up for anyone else that finds this and re-signs their commit history like I did: using the command in this Stack Overflow answer signed all but the first commit in my history, which prevented me from pushing any of the commits to a new remote repo that required commits be signed. I ended up needing to rebase on the initial commit already present in the remote repo.

**Reply by scj643** [reply]
I ran this to resign all commits and keep the commit-date 

`git rebase --committer-date-is-author-date --exec 'GIT_COMMITTER_DATE=$(git show -s --format=%ai) git commit --amend --no-edit' --root`

For some reason `committer-date-is-author-date` didn't seem to actually work

**Comment by Hazimehhassan** [comment]
@ShadyMedic ```

**Comment by wiktor-k** [comment]
FYI just in case anyone is looking for a reliable signed commits rewrite it's possible to set a committer on each commit via:

```
git filter-repo ...  --commit-callback '
    commit.committer_name = b"Signer Name"
    commit.committer_email = b"signer@example.com"
'
```

And then resign everything using your local key: `git filter-branch -f --commit-filter 'git commit-tree -S "$@";' -- --all`. It seems it works for all commits including the root one. Of course they now have "committer" set to the signing party, but this is expected.

:wave:
